### PR TITLE
Instantiation with templated function call

### PR DESCRIPTION
**Pull request by ansharlubis**
Continuing #2423, this PR allows instantiation of templated functions:
``` fortran
! given a templated function
pure function generic_sum {T, add, cast} (arr) result(res)
    requires operator_r(T, T, T, add)
    requires cast_r(T, cast)
    type(T), intent(in) :: arr(:)
    type(T) :: res
    integer :: n, i
    n = size(arr)
    res = cast(0)
    if (n > 0) then
        res = arr(1)
        do i=2,n
            res = add(res, arr(i))
        end do
    end if
end function

! the above function can be instantiated with
integer :: a_i(10), i, s_i
do i = 1, size(a_i)
    a_i(i) = i
    !a_r(i) = i
end do
s_i = generic_sum{integer, operator(+), cast_integer}(a_i)
```

This doesn't support overloading yet.

### CODE REVIEW

**Comment by ansharlubis** [comment]
Got multiple template function calls to work:
```
s_i = generic_sum{integer, operator(+), cast_integer}(a_i)
s_r = generic_sum{real, operator(+), cast_real}(a_r)
```
The templated function call generates a generic procedure `__templated_generic_sum` in which there are mapping to the instantiatied functions with concrete types.

File: grammar/AST.asdl [hunk 1]
```diff
@@ -200,7 +200,9 @@ expr
     | DefUnaryOp(string op, expr operand)
     | Compare(expr left, cmpop op, expr right)
     | FuncCallOrArray(identifier func, struct_member* member,
-        fnarg* args, keyword* keywords, fnarg* subargs)
+        fnarg* args, keyword* keywords, fnarg* subargs, use_symbol* temp_args)
+    --| TemplatedFuncCall(identifier func, struct_member* member,
+    --    fnarg* args, keyword* keywords, fnarg* subargs, use_symbol* temp_args)
```
Comment (certik) [inline grammar/AST.asdl:205]: I would remove the two commented out lines.

File: integration_tests/template_simple_01.f90 [hunk 2]
```diff
@@ -53,16 +53,14 @@ pure function generic_sum {T, add, cast} (arr) result(res)
...
             a_i(i) = i
             a_r(i) = i
         end do
-        s_i = generic_sum_integer(a_i)
-        s_r = generic_sum_real(a_r)
+        s_i = generic_sum{integer, operator(+), cast_integer}(a_i)
+        s_r = generic_sum{real, operator(+), cast_real}(a_r)
         print *, s_i
         print *, s_r
     end subroutine
```
Comment (certik) [inline integration_tests/template_simple_01.f90:66]: If we support both the explicit "instantiate" and "simple syntax", then I would keep both tests. In other words, create `template_simple_02.f90` for the simple instantiation syntax, and keep the old `template_smple_01.f90` unmodified.

**Review by certik — APPROVED** [review]
I left a few minor comments. Overall I think this looks good. Great job!

**Comment by certik** [comment]
CC @dacarnazzola, here is the simpler instantiation syntax.

**Review by certik — APPROVED** [review]
Looks great, thanks!

**Comment by certik** [comment]
I rebased it just in case, and set to auto-merge. Thanks!

**Pull request merged**